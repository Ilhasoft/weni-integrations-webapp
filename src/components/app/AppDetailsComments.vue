<template>
  <div class="app-details-comments">
    <div v-if="!editMode" class="app-details-comments__header">
      {{ $t('apps.details.comments.leave_a_comment') }}
    </div>
    <div v-else class="app-details-comments__header">
      {{ $t('apps.details.comments.edit_comment') + ' -&nbsp;' }}
      <div class="app-details-comments__header__cancel" @click="resetFields">
        {{ $t('general.Cancel') }}
      </div>
    </div>
    <unnnic-input
      ref="comment_input"
      class="app-details-comments__input"
      type="normal"
      iconRight="send-email-3-1"
      tooltipSideIconRight="right"
      size="md"
      :placeholder="$t('apps.details.comments.input_placeholder')"
      iconRightClickable
      @change="handleComment"
      v-model="currentComment"
    />
    <!-- TODO: Fix avatar and name when API is ready -->
    <unnnic-comment
      class="app-details-comments__comment"
      v-for="(comment, index) in commentsList"
      :key="index"
      :title="comment.owned ? $t('general.You') : comment.owner.first_name"
      :time="getCommentTime(comment.created_on)"
      :text="comment.content"
    >
      <Avatar
        v-if="!comment.owner.photo_url"
        :username="fullOwnerName(comment.owner)"
        slot="avatar"
      />
      <img
        class="app-details-comments__comment__avatar"
        v-else
        :src="comment.owner.photo_url"
        slot="avatar"
        alt=""
      />
      <unnnic-dropdown v-if="comment.owned" slot="actions">
        <unnnic-icon-svg slot="trigger" icon="navigation-menu-vertical-1" size="sm" />
        <unnnic-dropdown-item>
          <unnnic-button
            type="tertiary"
            iconLeft="pencil-write-1"
            :text="$t('apps.details.comments.edit_comment')"
            size="small"
            @click="handleUpdate(comment)"
          />
        </unnnic-dropdown-item>
        <unnnic-dropdown-item>
          <unnnic-button
            type="tertiary"
            iconLeft="delete-1"
            :text="$t('apps.details.comments.delete_comment')"
            size="small"
            @click="confirmDelete(comment.uuid)"
          />
        </unnnic-dropdown-item>
      </unnnic-dropdown>
    </unnnic-comment>

    <unnnic-modal
      ref="unnnic-remove-modal"
      :showModal="showRemoveModal"
      :text="$t('apps.details.comments.remove.title')"
      scheme="feedback-red"
      modal-icon="alert-circle-1"
      @close="toggleRemoveModal"
    >
      <span slot="message" v-html="$t('apps.details.comments.remove.description')"></span>
      <unnnic-button
        ref="unnnic-remove-modal-close-button"
        slot="options"
        type="tertiary"
        @click="toggleRemoveModal"
        >{{ $t('general.Cancel') }}</unnnic-button
      >
      <unnnic-button
        ref="unnnic-remove-modal-navigate-button"
        slot="options"
        type="primary"
        @click="handleDelete(currentRemovalUuid)"
        scheme="feedback-red"
      >
        {{ $t('apps.details.comments.remove.remove') }}
      </unnnic-button>
    </unnnic-modal>
  </div>
</template>

<script>
  import { app_type } from '@/stores/modules/appType/appType.store.js';
  import { mapActions, mapState } from 'pinia';
  import { comments_store } from '@/stores/modules/appType/comments/comments.store.js';
  import getRelativeTime from '../../utils/time.js';
  import { unnnicCallAlert } from '@weni/unnnic-system';
  import Avatar from 'vue-avatar';

  export default {
    name: 'AppDetailsComments',
    components: { Avatar },
    props: {
      appCode: {
        type: String,
        default: null,
      },
    },
    data() {
      return {
        currentComment: null,
        editMode: false,
        editCommentUuid: null,
        showRemoveModal: false,
        currentRemovalUuid: null,
      };
    },
    async mounted() {
      await this.listComments(this.appCode);
    },
    computed: {
      ...mapState(app_type, [
        'commentsList',
        'errorListComments',
        'errorCreateComment',
        'errorDeleteComment',
        'errorUpdateComment',
      ]),
    },
    methods: {
      ...mapActions(comments_store, [
        'listComments',
        'createComment',
        'deleteComment',
        'updateComment',
      ]),
      /* istanbul ignore next */
      getCommentTime(time) {
        const epoch = new Date(time).getTime();
        return `- ${getRelativeTime(epoch, this.$root.$i18n.locale)}`;
      },
      resetFields() {
        this.editMode = false;
        this.editCommentUuid = null;
        this.currentComment = null;
      },
      fullOwnerName(owner) {
        return owner.first_name + ' ' + owner.last_name;
      },
      toggleRemoveModal() {
        this.showRemoveModal = !this.showRemoveModal;
      },
      async handleComment() {
        if (this.currentComment !== null && this.currentComment.trim() !== '') {
          try {
            if (this.editMode) {
              await this.updateComment({
                code: this.appCode,
                commentUuid: this.editCommentUuid,
                payload: {
                  content: this.currentComment.trim(),
                },
              });

              if (this.errorUpdateComment) {
                throw new Error(this.errorUpdateComment);
              }
            } else {
              await this.createComment({
                code: this.appCode,
                payload: {
                  content: this.currentComment.trim(),
                },
              });

              if (this.errorCreateComment) {
                throw new Error(this.errorCreateComment);
              }
            }
            await this.listComments(this.appCode);

            if (this.errorListComments) {
              throw new Error(this.errorListComments);
            }
          } catch (err) {
            unnnicCallAlert({
              props: {
                text: this.$t('apps.details.status_error'),
                title: this.$t('general.error'),
                icon: 'check-circle-1-1',
                scheme: 'feedback-red',
                position: 'bottom-right',
                closeText: this.$t('general.Close'),
              },
              seconds: 3,
            });
          } finally {
            this.resetFields();
          }
        }
      },
      confirmDelete(commentUuid) {
        this.showRemoveModal = true;
        this.currentRemovalUuid = commentUuid;
      },
      async handleDelete(commentUuid) {
        try {
          await this.deleteComment({
            code: this.appCode,
            commentUuid,
          });

          if (this.errorDeleteComment) {
            throw new Error(this.errorDeleteComment);
          }

          this.showRemoveModal = false;
          this.currentRemovalUuid = null;
          await this.listComments(this.appCode);

          if (this.errorListComments) {
            throw new Error(this.errorListComments);
          }

          unnnicCallAlert({
            props: {
              text: this.$t('apps.details.comments.remove.status_text'),
              title: this.$t('general.success'),
              icon: 'check-circle-1-1',
              scheme: 'feedback-green',
              position: 'bottom-right',
              closeText: this.$t('general.Close'),
            },
            seconds: 3,
          });
        } catch (err) {
          unnnicCallAlert({
            props: {
              text: this.$t('apps.details.status_error'),
              title: this.$t('general.error'),
              icon: 'check-circle-1-1',
              scheme: 'feedback-red',
              position: 'bottom-right',
              closeText: this.$t('general.Close'),
            },
            seconds: 3,
          });
        }
      },
      handleUpdate(comment) {
        this.editMode = true;
        this.editCommentUuid = comment.uuid;
        this.currentComment = comment.content;

        this.$refs.comment_input.$el.children[0].children[0].focus();
      },
    },
  };
</script>

<style scoped lang="scss">
  .app-details-comments {
    &__header {
      display: flex;
      font-family: $unnnic-font-family-secondary;
      font-weight: $unnnic-font-weight-regular;
      font-size: $unnnic-font-size-body-gt;
      line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
      color: $unnnic-color-neutral-cloudy;
      margin-bottom: $unnnic-spacing-stack-xs;

      &__cancel {
        text-decoration: underline;
        cursor: pointer;
      }
    }

    &__input {
      margin-bottom: $unnnic-spacing-stack-md;
    }

    &__comment {
      margin-bottom: $unnnic-spacing-stack-md;

      &__avatar {
        width: 56px;
        object-fit: cover;
        border-radius: $unnnic-border-radius-pill;
      }
    }
  }
</style>
