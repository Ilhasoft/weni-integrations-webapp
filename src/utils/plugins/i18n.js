import Vue from 'vue';
import VueI18n from 'vue-i18n';

Vue.use(VueI18n);

import en from '../../locales/en.json';
import pt_br from '../../locales/pt_br.json';
import es_es from '../../locales/es_es.json';

const languages = {
  'en-us': en,
  'pt-br': pt_br,
  es: es_es,
};

const messages = Object.assign(languages);

const i18n = new VueI18n({
  locale: 'en-us',
  fallbackLocale: 'en-us',
  messages,
});

export default i18n;
