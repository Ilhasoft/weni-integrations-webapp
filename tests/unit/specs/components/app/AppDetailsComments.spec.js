import { unnnicCallAlert as mockUnnnicCallAlert } from '@weni/unnnic-system';

jest.mock('@weni/unnnic-system', () => ({
  ...jest.requireActual('@weni/unnnic-system'),
  unnnicCallAlert: jest.fn(),
}));

import { mount, createLocalVue } from '@vue/test-utils';
import AppDetailsComments from '@/components/app/AppDetailsComments';
import i18n from '@/utils/plugins/i18n';
import VueRouter from 'vue-router';
import Vuex from 'vuex';

const localVue = createLocalVue();
localVue.use(VueRouter);
localVue.use(Vuex);

describe('AppDetailsComments.vue', () => {
  let wrapper;
  let store;
  let actions;

  beforeEach(() => {
    actions = {
      listComments: jest.fn(() => {
        return { data: [] };
      }),
      createComment: jest.fn(),
      deleteComment: jest.fn(),
      updateComment: jest.fn(),
    };

    store = new Vuex.Store({
      actions,
    });

    wrapper = mount(AppDetailsComments, {
      localVue,
      store,
      i18n,
      mocks: {
        $t: () => 'some specific text',
      },
      stubs: {
        UnnnicButton: true,
        UnnnicComment: true,
        UnnnicDropdown: true,
        UnnnicDropdownItem: true,
        UnnnicIconSvg: true,
      },
      propsData: {
        appCode: 'code',
      },
      attachTo: document.body,
    });

    jest.clearAllMocks();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should be rendered properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('fullOwnerName()', () => {
    it('should return full name', () => {
      const owner = {
        first_name: 'Jane',
        last_name: 'Doe',
      };
      const fullName = wrapper.vm.fullOwnerName(owner);
      expect(fullName).toEqual('Jane Doe');
    });
  });

  describe('toggleRemoveModal()', () => {
    it('should toggle', () => {
      expect(wrapper.vm.showRemoveModal).toBeFalsy();
      wrapper.vm.toggleRemoveModal();
      expect(wrapper.vm.showRemoveModal).toBeTruthy();
    });
  });

  describe('confirmDelete()', () => {
    it('should set showRemoveModal to true', () => {
      const uuid = '123';
      expect(wrapper.vm.showRemoveModal).toBeFalsy();
      wrapper.vm.confirmDelete(uuid);
      expect(wrapper.vm.showRemoveModal).toBeTruthy();
    });

    it('should set currentRemovalUuid', () => {
      const uuid = '123';
      expect(wrapper.vm.currentRemovalUuid).not.toEqual(uuid);
      wrapper.vm.confirmDelete(uuid);
      expect(wrapper.vm.currentRemovalUuid).toEqual(uuid);
    });
  });

  describe('handleComment()', () => {
    const comment = {
      uuid: '123',
      text: 'comment ', // whitespace on end to test trim()
    };
    beforeEach(() => {
      wrapper.vm.currentComment = comment.text;
    });

    it('should do nothing if currentComment is empty', async () => {
      wrapper.vm.currentComment = '   ';
      const spy = spyOn(wrapper.vm, 'resetFields');
      expect(spy).not.toHaveBeenCalled();

      await wrapper.vm.handleComment();

      expect(spy).not.toHaveBeenCalled();
    });

    it('should call fetchComments()', async () => {
      const spy = spyOn(wrapper.vm, 'fetchComments');
      expect(spy).not.toHaveBeenCalled();

      await wrapper.vm.handleComment();

      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(wrapper.vm.appCode);
    });

    it('should call unnnicCallAlert on error', async () => {
      spyOn(wrapper.vm, 'fetchComments').and.throwError(new Error('error fetching'));
      expect(mockUnnnicCallAlert).not.toHaveBeenCalled();
      await wrapper.vm.handleComment();
      expect(mockUnnnicCallAlert).toHaveBeenCalledTimes(1);
    });

    it('should call resetFields() if currentComment is valid', async () => {
      const spy = spyOn(wrapper.vm, 'resetFields');
      expect(spy).not.toHaveBeenCalled();

      await wrapper.vm.handleComment();

      expect(spy).toHaveBeenCalledTimes(1);
    });

    describe('ADD mode', () => {
      beforeEach(() => {
        wrapper.vm.editMode = false;
      });

      it('should call createComment()', async () => {
        expect(actions.createComment).not.toHaveBeenCalled();
        await wrapper.vm.handleComment();
        expect(actions.createComment).toHaveBeenCalledTimes(1);
        expect(actions.createComment).toHaveBeenCalledWith(expect.any(Object), {
          code: wrapper.vm.appCode,
          payload: {
            content: comment.text.trim(),
          },
        });
      });
    });

    describe('EDIT mode', () => {
      beforeEach(() => {
        wrapper.vm.editMode = true;
        wrapper.vm.editCommentUuid = comment.uuid;
      });

      it('should call updateComment()', async () => {
        expect(actions.updateComment).not.toHaveBeenCalled();
        await wrapper.vm.handleComment();
        expect(actions.updateComment).toHaveBeenCalledTimes(1);
        expect(actions.updateComment).toHaveBeenCalledWith(expect.any(Object), {
          code: wrapper.vm.appCode,
          commentUuid: comment.uuid,
          payload: {
            content: comment.text.trim(),
          },
        });
      });
    });
  });

  describe('handleDelete()', () => {
    const commentUuid = 123;
    it('should call deleteComment()', async () => {
      expect(actions.deleteComment).not.toHaveBeenCalled();

      await wrapper.vm.handleDelete(commentUuid);

      expect(actions.deleteComment).toHaveBeenCalledTimes(1);
      expect(actions.deleteComment).toHaveBeenCalledWith(expect.any(Object), {
        code: wrapper.vm.appCode,
        commentUuid,
      });
    });

    it('should call fetchComments()', async () => {
      const spy = spyOn(wrapper.vm, 'fetchComments');
      expect(spy).not.toHaveBeenCalled();

      await wrapper.vm.handleDelete(commentUuid);

      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(wrapper.vm.appCode);
    });

    it('should call unnnicCallAlert() on error', async () => {
      spyOn(wrapper.vm, 'fetchComments').and.throwError(new Error('error fetching'));
      expect(mockUnnnicCallAlert).not.toHaveBeenCalled();
      await wrapper.vm.handleDelete(commentUuid);
      expect(mockUnnnicCallAlert).toHaveBeenCalledTimes(1);
    });
  });

  describe('handleUpdate()', () => {
    const comment = {
      uuid: '123',
      content: 'text',
    };

    it('should set on edit mode', () => {
      expect(wrapper.vm.editMode).toBeFalsy();
      wrapper.vm.handleUpdate(comment);
      expect(wrapper.vm.editMode).toBeTruthy();
    });

    it('should set comment data to edit', () => {
      expect(wrapper.vm.editCommentUuid).toBeNull();
      expect(wrapper.vm.currentComment).toBeNull();
      wrapper.vm.handleUpdate(comment);
      expect(wrapper.vm.editCommentUuid).toEqual(comment.uuid);
      expect(wrapper.vm.currentComment).toEqual(comment.content);
    });

    it('should set focus on input', () => {
      const input = wrapper.findComponent({ ref: 'comment_input' }).element.children[0].children[0];
      expect(input).not.toBe(document.activeElement);
      wrapper.vm.handleUpdate(comment);
      expect(input).toBe(document.activeElement);
    });
  });
});
