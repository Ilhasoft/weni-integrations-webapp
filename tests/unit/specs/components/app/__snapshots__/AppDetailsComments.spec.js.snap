// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppDetailsComments.vue should be rendered properly 1`] = `
<div class="app-details-comments">
  <div class="app-details-comments__header">
    some specific text
  </div>
  <div class="app-details-comments__input unnnic-form" placeholder="some specific text">
    <!---->
    <div data-v-ea26bad8="" class="unnnic-form-input unnnic-form has-icon" v-bind="[object Object]" mask=""><input data-v-ea26bad8="" size="0" type="text" v-bind="[object Object]" placeholder="some specific text" class="unnnic-form__input unnnic-form__input--size-md has-icon-right">
      <!---->
      <unnniciconsvg-stub data-v-ea26bad8="" icon="send-email-3-1" clickable="true" size="sm" scheme="neutral-cloudy" class="icon-right"></unnniciconsvg-stub>
    </div>
    <!---->
  </div>
  <!---->
</div>
`;
